libyul/AsmAnalysis.cpp

```diff
@@ -669,8 +659,16 @@ void AsmAnalyzer::warnOnInstructions(solidity::Instruction _instr, SourceLocatio
 		_instr == solidity::Instruction::SHL ||
 		_instr == solidity::Instruction::SHR ||
 		_instr == solidity::Instruction::SAR ||
-		_instr == solidity::Instruction::CREATE2
+		_instr == solidity::Instruction::CREATE2 ||
+		_instr == solidity::Instruction::EXTCODEHASH
 	) && !m_evmVersion.hasBitwiseShifting())
+	{
+		solAssert(
+			!m_evmVersion.hasBitwiseShifting() &&
+			!m_evmVersion.hasCreate2() &&
+			!m_evmVersion.hasExtCodeHash(),
```

Comment:
Instead of this assert we could also reorg the if statement:
```
((shl || shr || sar) && !hasBitwise) ||
(create2 && !hasCreate2) ||
(extcodehash && !hasExtCodeHash)
```

Reply:
I mean yeah, this should be more specific. It should be either `if (instr == shr || ... || == extcodehash) solAssert(m_evmVersion <= Constantinople, ...)` or split into individual cases. I would prefer using actual IFs, though.

Reply:
What do you think about a helper like this?

```
        auto warningForVM = [=](string const& vmKindMessage) {
                m_errorReporter.warning(
                        _location,
                        "The \"" +
                        boost::to_lower_copy(instructionInfo(_instr).name)
                        + "\" instruction is " +
                        vmKindMessage +
                        " VMs. " +
                        "You are currently compiling for \"" +
                        m_evmVersion.name() +
                        "\", where it will be interpreted as an invalid instruction."
                );
        };

        if ((
                _instr == solidity::Instruction::RETURNDATACOPY ||
                _instr == solidity::Instruction::RETURNDATASIZE ||
                _instr == solidity::Instruction::STATICCALL
        ) && !m_evmVersion.supportsReturndata())
                warningForVM("only available for Byzantium-compatible");
```

Reply:
looks good!

Reply:
Updated, not sure if it is better.